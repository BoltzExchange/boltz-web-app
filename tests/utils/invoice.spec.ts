import bolt11 from "bolt11";
import { describe, expect, test } from "vitest";

import {
    decodeInvoice,
    getExpiryEtaHours,
    isLnurl,
    trimLightningPrefix,
} from "../../src/utils/invoice";

describe("invoice", () => {
    test.each`
        data                                                                             | expected
        ${"m@lnurl.some.domain"}                                                         | ${true}
        ${"LNURL1DP68GURN8GHJ7MRWW4EXCTNDD93KSCT9DSCNQVF39ESHGTMPWP5J7MRWW4EXCUQGY84ZH"} | ${true}
        ${"lnurl.some.domain"}                                                           | ${false}
        ${"LNURL1DP6fasdklfjasdf"}                                                       | ${false}
    `(
        "should determine if $data is lnurl ($expected)",
        ({ data, expected }) => {
            expect(isLnurl(data)).toEqual(expected);
        },
    );

    test("should trim lightning: prefix of invoices", () => {
        const invoice = "lnbcrt4986620n1pjgkj07pp5zl";

        expect(trimLightningPrefix(invoice)).toEqual(invoice);
        expect(trimLightningPrefix(`lightning:${invoice}`)).toEqual(invoice);
        expect(trimLightningPrefix(`LIGHTNING:${invoice}`)).toEqual(invoice);
    });

    describe("decode invoices", () => {
        test.each`
            network      | invoice
            ${"regtest"} | ${"lnbcrt623210n1pj8hfdspp5mhcxq3qgzn779zs0c02na32henclzt55uga68kck6tknyw0y59qsdqqcqzzsxqyz5vqsp54wll9s5jphgcjqzpnamqeszvfdz937pjels2cqr84pltjsqv2asq9qyyssq49028nqec7uz5vk73peg5a4fkxhltw90kkmupfradjp0sus6g5zxs6njedk8ml3qgdls3dfjfvd7z3py5qgst9fnzz5pwcr5564sf6sqtrlfzz"}
            ${"testnet"} | ${"lntb4573450n1pj8hfnmsp58erxc4m9u09frqkalhw5udgvghvm27wewl99d7z3hjftgwtt234qpp572p5y0tplt70txw35kzypsef4mg2pwp5u0ej9hx8tse6f2rcvrjsdq5g9kxy7fqd9h8vmmfvdjsxqyjw5qcqp2rzjq0cxp9fmaadhwlw80ez2lgu9n5pzlsd803238r0tyv4dwf27s6wqqfggesqqqfqqqyqqqqlgqqqqqqgq9q9qyysgqw4msvmfgakcmxkglwnj7qgp6hlupefstyzhkld0uxlx3gdncnzw385c5qy6ng2qh59rtttktjzy8l43gzv3n9u6du64z2xu0mdz377splwf2qy"}
            ${"mainnet"} | ${"lnbc678450n1pj8hf4kpp5kxh4x93kvxt43q0k0q6t3fp6gfhgusqxsajj6lcexsrg4lzm7rrqdq5g9kxy7fqd9h8vmmfvdjscqzzsxqyz5vqsp5n4rzwr2lzw68082ws4tjjerp2t5eluny75xx54jr530x073tvvzs9qyyssq3f43e2mzqx07zzt529ux480nj00908p3u5qdwhyuk3qrcepaqsjxqjhcnfde4ta74c3dkxkhwscxfhdm5v0y7qh7np22v9xc220taacqjanm3m"}
        `("should decode $network invoices", ({ invoice }) => {
            expect(decodeInvoice(invoice)).toMatchSnapshot();
        });
    });

    describe("Expiry Eta hours", () => {
        test.each`
            invoice
            ${"lnbcrt623210n1pj8hfdspp5mhcxq3qgzn779zs0c02na32henclzt55uga68kck6tknyw0y59qsdqqcqzzsxqyz5vqsp54wll9s5jphgcjqzpnamqeszvfdz937pjels2cqr84pltjsqv2asq9qyyssq49028nqec7uz5vk73peg5a4fkxhltw90kkmupfradjp0sus6g5zxs6njedk8ml3qgdls3dfjfvd7z3py5qgst9fnzz5pwcr5564sf6sqtrlfzz"}
        `("should decode eta time expired", ({ invoice }) => {
            expect(getExpiryEtaHours(invoice)).toEqual(0);
        });

        test.each`
            delta
            ${12}
            ${24}
        `("expiry eta in $delta hours", ({ delta }) => {
            const now = +new Date() / 1000;
            const encoded = bolt11.encode({
                satoshis: 2000,
                timestamp: now,
                tags: [
                    {
                        tagName: "payment_hash",
                        data: "0001020304050607080900010203040506070809000102030405060708090102",
                    },
                    {
                        tagName: "expire_time",
                        data: delta * 60 * 60,
                    },
                ],
            });
            // sign takes the encoded object and the private key as arguments
            const privateKeyHex =
                "e126f68f7eafcc8b74f54d269fe206be715000f94dac067d1c04a8ca3b2db734";
            const signed = bolt11.sign(encoded, privateKeyHex);
            expect(getExpiryEtaHours(signed.paymentRequest)).toEqual(delta);
        });
    });
});
