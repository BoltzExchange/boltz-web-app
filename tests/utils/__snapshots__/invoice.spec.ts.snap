// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`invoice decode invoices should decode mainnet invoices 1`] = `
{
  "expiry": 1685911606,
  "preimageHash": "b1af53163661975881f67834b8a43a426e8e400687652d7f1934068afc5bf0c6",
  "satoshis": 67845,
}
`;

exports[`invoice decode invoices should decode regtest invoices 1`] = `
{
  "expiry": 1685911344,
  "preimageHash": "ddf060440814fde28a0fc3d53ec557ccf1f12e94e23ba3db16d2ed3239e4a141",
  "satoshis": 62321,
}
`;

exports[`invoice decode invoices should decode testnet invoices 1`] = `
{
  "expiry": 1686429947,
  "preimageHash": "f283423d61fafcf599d1a58440c329aed0a0b834e3f322dcc75c33a4a87860e5",
  "satoshis": 457345,
}
`;
